### PR TITLE
Assorted QoL changes

### PR DESCRIPTION
**Pull request by kiwiz**
**Implemented**

* Fixes physical keyboard repeat
    * This was broken by my previous PR
* Handle screen input rotation
    * Not sure if gestures are handled correctly, but I'm not sure if any gestures are actually in use?
* Clear garbage on screen edges 
* Support for F1 - F12 keys

**Todo**

* ~Dynamically detect when keyboard is attached~ Tabling for now
* ~Config support~
* ~BUG: Keyboard visibility state is broken~
    * Repro: Trigger the keyboard hide/show code 3 times.
* ~Fix F1 - F12 support. May need to update the terminfo for this to work properly? Help with this would be appreciated~

### CODE REVIEW

File: apps/yaft/keyboard.cpp [hunk 1]
```diff
@@ -402,16 +468,26 @@ Keyboard::initKeyMap() {
...
+  // calculate valid screen region
+  startHeight = term->height - (term->isLandscape ?
+      0 :
+      (hidden ? hidden_keyboard_height : keyboard_height)
+   ) - 1;
+
+  this->screenRect = Rect{
+    { term->marginLeft, term->marginTop },
+    { term->width - term->marginLeft - 1, startHeight - term->marginTop }
+  };
```
Comment (timower) [inline apps/yaft/keyboard.cpp:480]: Are you using clang-format?

File: apps/yaft/terminal.cpp [hunk 2]
```diff
@@ -381,17 +381,9 @@ bool
...
   extern const uint32_t color_list[COLORS]; /* global */
 
-  term->width = width;
-  term->height = height;
-  term->isLandscape = isLandscape;
-
-  /* 1 px margin on edges */
-  term->cols = (term->width - 2) / CELL_WIDTH;
-  term->lines = (term->height - 2) / CELL_HEIGHT;
+  term_resize(term, width, height, isLandscape, false);
```
Comment (timower) [inline apps/yaft/terminal.cpp:384]:
```suggestion
  term_resize(term, width, height, isLandscape, /* report */ false);
```

**Review by timower — CHANGES_REQUESTED** [review]
Small nits mostly

**Comment by timower** [comment]
> Not sure if gestures are handled correctly, but I'm not sure if any gestures are actually in use?
Two finger scroll should work, but it is kind of flaky.

**Comment by timower** [comment]
> I noticed that hovering over the screen with the pen triggers inputs. IMO, only physically tapping the screen should report a mouse input.

The idea was to allow mouse moves when hovering. That way you could drag something by touching the screen while moving the pen. Though I don't use the mouse integration that much, so I'm fine with either implementation or a config option.

File: apps/yaft/keyboard.cpp [hunk 3]
```diff
@@ -617,7 +689,7 @@ struct event_traits<PenEvent> {
 };
 
 void
-initMouseBuf(std::array<char, 6>& buf, Point loc) {
+initMouseBuf(terminal_t* term, std::array<char, 6>& buf, Point loc) {
```
[the remark this answers is not in the document]
Comment (kiwiz) [reply apps/yaft/keyboard.cpp:692]: Good catch.

**Comment by kiwiz** [comment]
> The idea was to allow mouse moves when hovering. That way you could drag something by touching the screen while moving the pen. Though I don't use the mouse integration that much, so I'm fine with either implementation or a config option.

Ah, I see. In that case, I'll see if I can require a physical touch before events start getting fired (if that's not already the case).

File: apps/yaft/keyboard.cpp (hunk 1 above)
Comment (kiwiz) [reply apps/yaft/keyboard.cpp]: I am now!

**Comment by timower** [comment]
> BUG: Keyboard visibility state is broken
Just fixed this, along with pogo detection in [yaft_auto](https://github.com/timower/rM2-stuff/tree/yaft_auto_landscape), I'll get my folio tomorrow to verify if it works.

**Review by timower — APPROVED** [review]
One more nit, but then I'll merge

**Comment by kiwiz** [comment]
Sweet, hope you enjoy it!

I'll open up a separate PR for the config support (unless if you'd like to take a stab at it) and continue investigating the F# keys.

File: apps/yaft/keyboard.cpp [hunk 4]
```diff
@@ -288,6 +300,23 @@ getKeyCodeStr(int scancode, bool shift, bool alt, bool ctrl, bool appCursor) {
...
+
+  constexpr auto write_hi_f_code = [](char a, char b) {
+    buf[0] = esc_char;
+    buf[1] = '[';
+    buf[2] = a;
+    buf[3] = b;
+    buf[4] = '~';
+    buf[5] = 0;
+  };
+
```
Comment (timower) [inline apps/yaft/keyboard.cpp:319]: From https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-PC-Style-Function-Keys
```suggestion
  constexpr auto write_ss3 = [](char code) {
    buf[0] = esc_char;
    buf[1] = 'O';
    buf[3] = code;
    buf[4] = 0;
  };

  constexpr auto write_hi_vt_code = [](char a, char b) {
    buf[0] = esc_char;
    buf[1] = '[';
    buf[2] = a;
    buf[3] = b;
    buf[4] = '~';
    buf[5] = 0;
  };

```

File: apps/yaft/keyboard.cpp [hunk 5]
```diff
@@ -343,6 +372,43 @@ getKeyCodeStr(int scancode, bool shift, bool alt, bool ctrl, bool appCursor) {
...
+        return buf.data();
+      case F2:
+        write_f_code('B');
+        return buf.data();
+      case F3:
+        write_f_code('C');
+        return buf.data();
+      case F4:
+        write_f_code('D');
+        return buf.data();
```
Comment (timower) [inline apps/yaft/keyboard.cpp:387]: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-PC-Style-Function-Keys
```suggestion
      case F1:
        write_ss3('p');
        return buf.data();
      case F2:
        write_ss3('Q');
        return buf.data();
      case F3:
        write_ss3('R');
        return buf.data();
      case F4:
        write_ss3('S');
        return buf.data();
```

File: apps/yaft/keyboard.cpp [hunk 6]
```diff
@@ -343,6 +372,43 @@ getKeyCodeStr(int scancode, bool shift, bool alt, bool ctrl, bool appCursor) {
...
+      case F10:
+        write_hi_f_code('2', '1');
+        return buf.data();
+      case F11:
+        write_hi_f_code('2', '3');
+        return buf.data();
+      case F12:
+        write_hi_f_code('2', '4');
+        return buf.data();
```
Comment (timower) [inline apps/yaft/keyboard.cpp:411]:
```suggestion
      case F5:
        write_hi_vt_code('1', '5');
        return buf.data();
      case F6:
        write_hi_vt_code('1', '7');
        return buf.data();
      case F7:
        write_hi_vt_code('1', '8');
        return buf.data();
      case F8:
        write_hi_vt_code('1', '9');
        return buf.data();
      case F9:
        write_hi_vt_code('2', '0');
        return buf.data();
      case F10:
        write_hi_vt_code('2', '1');
        return buf.data();
      case F11:
        write_hi_vt_code('2', '3');
        return buf.data();
      case F12:
        write_hi_vt_code('2', '4');
        return buf.data();
```

**Review by timower — CHANGES_REQUESTED** [review]
This should hopefully fix the F keys

File: apps/yaft/keyboard.cpp (hunk 6 above)
Comment (kiwiz) [reply apps/yaft/keyboard.cpp:410]: Testing.

**Comment by kiwiz** [comment]
> This should hopefully fix the F keys

Looks like F6 - F12 work (in `htop`, at least)

**Comment by kiwiz** [comment]
With the changes above, F5 - F12 work.

**Comment by kiwiz** [comment]
Ah, indices were off on the `write_ss3` function. All F# keys work now!

**Review by timower — APPROVED** [review]

**Pull request merged**